src/rpc/evo.cpp

```diff
@@ -746,41 +741,59 @@ static UniValue protx_register_common_wrapper(const JSONRPCRequest& request,
+            if (!wallet->IsLockedCoin(ptx.collateralOutpoint.hash, ptx.collateralOutpoint.n)) {
+                wallet->LockCoin(ptx.collateralOutpoint);
+                unlockOnError = true;
+            }
         }
-
-        if (isPrepareRegister) {
-            // external signing with collateral key
-            ptx.vchSig.clear();
-            SetTxPayload(tx, ptx);
-
-            UniValue ret(UniValue::VOBJ);
-            ret.pushKV("tx", EncodeHexTx(CTransaction(tx)));
-            ret.pushKV("collateralAddress", EncodeDestination(txDest));
-            ret.pushKV("signMessage", ptx.MakeSignString());
-            return ret;
-        } else {
-            // lets prove we own the collateral
-            LegacyScriptPubKeyMan* spk_man = wallet->GetLegacyScriptPubKeyMan();
-            if (!spk_man) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "This type of wallet does not support this command");
+        try {
+            FundSpecialTx(wallet.get(), tx, ptx, fundDest);
+            UpdateSpecialTxInputsHash(tx, ptx);
+            Coin coin;
+            if (!GetUTXOCoin(ptx.collateralOutpoint, coin)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral not found: %s", ptx.collateralOutpoint.ToStringShort()));
+            }
+            CTxDestination txDest;
+            ExtractDestination(coin.out.scriptPubKey, txDest);
+            const PKHash* pkhash = std::get_if<PKHash>(&txDest);
+            if (!pkhash) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral type not supported: %s", ptx.collateralOutpoint.ToStringShort()));
             }
 
-            CKey key;
-            if (!spk_man->GetKey(ToKeyID(*pkhash), key)) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+            if (isPrepareRegister) {
+                // external signing with collateral key
+                ptx.vchSig.clear();
+                SetTxPayload(tx, ptx);
+
+                UniValue ret(UniValue::VOBJ);
+                ret.pushKV("tx", EncodeHexTx(CTransaction(tx)));
+                ret.pushKV("collateralAddress", EncodeDestination(txDest));
+                ret.pushKV("signMessage", ptx.MakeSignString());
+                return ret;
+            } else {
+                // lets prove we own the collateral
+                LegacyScriptPubKeyMan* spk_man = wallet->GetLegacyScriptPubKeyMan();
+                if (!spk_man) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "This type of wallet does not support this command");
+                }
+
+                CKey key;
+                if (!spk_man->GetKey(ToKeyID(*pkhash), key)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
+                }
+                SignSpecialTxPayloadByString(tx, ptx, key);
```

Comment:
nit
```suggestion
                if (CKey key; !spk_man->GetKey(ToKeyID(*pkhash), key)) {
                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
                } else {
                    SignSpecialTxPayloadByString(tx, ptx, key);
                }
```

Reply:
I disagree

Reply:
good thing it was a nit :P

Reply:
yeah, I'm not sure which version I like more actually 🙈

Reply:
I'd prefer original one, bcs throw is throw, else looks ugly for throw

Reply:
That's also an old code, there's just changed amount of whitespace but code is same byte-to-byte

Reply:
we could maybe write it like that to make it less ugly :) while still keeping `key` inside of the `if` scope
```suggestion
                if (CKey key; spk_man->GetKey(ToKeyID(*pkhash), key)) {
                    SignSpecialTxPayloadByString(tx, ptx, key);
                } else {
                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("collateral key not in wallet: %s", EncodeDestination(txDest)));
                }
```

Reply:
> That's also an old code, there's just changed amount of whitespace but code is same byte-to-byte

Good point, makes it easier to review with `?w=1` indeed

Reply:
for the moment I'm leaving it as it is since I did not modify that lines